File: Java/MysticDave/TextureSheet.java
/*
========================================================================

	A texture sheet of same-sized textures

========================================================================
*/
package MysticDave;

import Managers.ResourceManager;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.Image;


public class TextureSheet {
    
    private Image textureImage;
    private int textureWidth;  // number of pixels each texture is wide
    private int textureHeight; // number of pixels each texture is tall
    private int numWidth;      // number of textures across image
    private int numHeight;     // number of textures down image
    
    public TextureSheet( String name, int width, int height ) {
        this.textureWidth  = width;
	this.textureHeight = height;

	this.textureImage = (ResourceManager.GetInstance()).GetBitmap( name );
	int w = textureImage.getWidth();
	int h = textureImage.getHeight();

	this.numWidth  = w / this.textureWidth;
	this.numHeight = h / this.textureHeight;
    }
    
    public Image GetTexture( int id ) {
        int sx = id % numWidth;
	int sy = id / numWidth;

	return textureImage.getSubImage( sx*textureWidth, sy*textureWidth, textureWidth, textureHeight );
    }
    
    public void RenderTexture( Graphics g, int id, float x, float y ) {
        int sx = (id % numWidth) * textureWidth;
	int sy = (id / numWidth) * textureHeight;
        
        g.drawImage( textureImage, x, y, x + textureWidth, y + textureHeight, sx, sy, sx+textureWidth, sy+textureHeight );
    }
    
    public void RenderTextureByCenter( Graphics g, int id, float x, float y ) {
        int sx = (id % numWidth) * textureWidth;
	int sy = (id / numWidth) * textureHeight;

        g.drawImage( textureImage, x - textureWidth/2, y - textureHeight/2, x + textureWidth/2, y + textureHeight/2, sx, sy, sx+textureWidth, sy+textureHeight );
    }
    
}

File: CPP/TextureSheet.cpp
/*
========================================================================

	A texture sheet of same-sized textures

    Author: Tim Wheeler
    Contact: timwheeleronline@gmail.com

========================================================================
*/

#include "TextureSheet.h"
#include "ResourceManager.h"
#include "LogManager.h"

using namespace MysticDave;

TextureSheet::TextureSheet( const char * name, int width, int height ) {

	TextureSheet::textureWidth  = width;
	TextureSheet::textureHeight = height;

	TextureSheet::textureImage = (ResourceManager::GetInstance()).GetBitmap( name );
	int w = al_get_bitmap_width(textureImage);
	int h = al_get_bitmap_height(textureImage);

	TextureSheet::numWidth  = w / TextureSheet::textureWidth;
	TextureSheet::numHeight = h / TextureSheet::textureHeight;
}

TextureSheet::~TextureSheet() {
	// do nothing
}

ALLEGRO_BITMAP * TextureSheet::GetTexture( int id ) {
	int sx = id % numWidth;
	int sy = id / numWidth;

	return al_create_sub_bitmap( textureImage, sx*textureWidth, sy*textureHeight, textureWidth, textureHeight );
}

//renders via top-left corner of texture
void TextureSheet::RenderTexture( int id, float x, float y ) {
	int sx = id % numWidth;
	int sy = id / numWidth;

	al_draw_bitmap_region( textureImage, sx*textureWidth, sy*textureHeight, textureWidth, textureHeight, x, y, 0 );
}

void TextureSheet::RenderTextureByCenter( int id, float x, float y ) {

	int sx = id % numHeight;
	int sy = id / numHeight;

	al_draw_bitmap_region( textureImage, sx*textureWidth, sy*textureHeight, textureWidth, textureHeight, x - textureWidth/2, y - textureHeight/2, 0 );

}